NotificationHubs/src/com/windowsazure/messaging/NotificationHubsExceptionFactory.java

```diff
@@ -0,0 +1,72 @@
+//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//----------------------------------------------------------------
+
+package com.windowsazure.messaging;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.time.Duration;
+import java.util.Optional;
+
+public class NotificationHubsExceptionFactory {
+
+    static boolean isTransientStatusCode(int httpStatusCode) {
+        return httpStatusCode == 403 || httpStatusCode == 408 || httpStatusCode == 429 || httpStatusCode == 500
+                || httpStatusCode == 503 || httpStatusCode == 504;
+    }
+
+    static Optional<Duration> parseRetryAfter(HttpResponse response) {
+        Header retryAfter = response.getFirstHeader(HttpHeaders.RETRY_AFTER);
+        if (retryAfter == null) {
+            return Optional.empty();
+        }
+        String retryAfterValue = retryAfter.getValue();
+        if (retryAfterValue.equals("")) {
+            return Optional.empty();
+        }
+
+        try {
+            long retryAfterSeconds = Long.parseLong(retryAfterValue);
+            return Optional.of(Duration.ofSeconds(retryAfterSeconds));
+        } catch (NumberFormatException e) {
+            throw new UnsupportedOperationException(
+                    String.format("\"%s\" must be an integer number of seconds", retryAfterValue));
+        }
+    }
+
+    static String getErrorString(HttpResponse response) throws IllegalStateException, IOException {
+        StringWriter writer = new StringWriter();
+        IOUtils.copy(response.getEntity().getContent(), writer, "UTF-8");
+        String body = writer.toString();
+        return "Error: " + response.getStatusLine() + " - " + body;
+    }
+
+    public static NotificationHubsException createNotificationHubException(HttpResponse response, int httpStatusCode)
```

Comment:
Think it would be worth building the same Notification heirarchy that exists in [the Android library](https://github.com/Azure/azure-notificationhubs-android/tree/master/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs)? There are only a few classes at the moment: `AuthorizationException`, `ClientException`, `NotificationHubException`, and `ServerException`.

Reply:
We cannot introduce breaking changes at this point, which would be to keep the existing NotificationHubsException and nothing more since we're introducing the HTTP status code, whether the error is transient, and the retry-after duration which should be sufficient for any client.